src/lib.rs

```diff
@@ -0,0 +1,185 @@
+            serializer_map
+                .serialize_entry("timestamp", &timestamp)
+                .unwrap();
+        }
+
+        serializer_map
+            .serialize_entry("level", &meta.level().as_serde())
+            .unwrap();
+
+        let mut visitor = tracing_serde::SerdeMapVisitor::new(serializer_map);
+        event.record(&mut visitor);
+        let mut serializer_map = visitor.take_serializer().unwrap();
+
+        if let Some(scope) = ctx.event_scope() {
+            for (index, span) in scope.enumerate() {
+                if index == 0 {
+                    serializer_map.serialize_entry("span", span.name()).unwrap();
+                }
+
+                let ext = span.extensions();
+                if let Some(data) = ext.get::<FormattedFields<N>>() {
+                    if let serde_json::Value::Object(fields) =
+                        serde_json::from_str::<serde_json::Value>(data).unwrap()
+                    {
+                        for field in fields {
+                            serializer_map.serialize_entry(&field.0, &field.1).unwrap();
+                        }
+                    }
+                }
+            }
+        }
+
+        serializer_map.end().unwrap();
+
+        writer.write_str(std::str::from_utf8(&s).unwrap()).unwrap();
+        writeln!(writer)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::{
+        io,
+        sync::{Arc, Mutex},
+    };
+
+    use tracing::{dispatcher, info};
+    use tracing_subscriber::{fmt::format::JsonFields, Layer, Registry};
+
+    use super::*;
+
+    #[derive(Debug, Clone)]
+    struct TestWriter {
+        data: Arc<Mutex<Vec<u8>>>,
+    }
+
+    impl TestWriter {
+        fn new() -> Self {
+            Self {
+                data: Arc::new(Mutex::new(Vec::new())),
+            }
+        }
+    }
+
+    impl io::Write for TestWriter {
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            self.data.lock().unwrap().write(buf)
+        }
+
+        fn flush(&mut self) -> io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[tokio::test]
+    async fn should_write_a_log() {
+        let writer = TestWriter::new();
+
+        let log_to_file = {
+            let writer = writer.clone();
+            tracing_subscriber::fmt::layer()
+                .event_format(SolinkJsonFormat::new().with_timestamp(false))
+                .fmt_fields(JsonFields::default())
+                .with_writer(move || writer.clone())
+        };
+
+        let subscriber = log_to_file.with_subscriber(Registry::default());
+        let dispatch = dispatcher::Dispatch::new(subscriber);
+        dispatcher::with_default(&dispatch, || {
+            let span1 = tracing::info_span!("parent", x = 7);
+            let span2 = tracing::info_span!(parent: &span1, "child", y = 9);
+
+            let _s1 = span1.enter();
+            let _s2 = span2.enter();
+
+            info!(z = 10, "Test")
```

Comment:
If you add a `y` field here like:

```rust
info!(z = 10, y = 42, "Test")
```

then it serializes `y` twice:

```
"{\"level\":\"INFO\",\"message\":\"Test\",\"z\":10,\"y\":42,\"span\":\"child\",\"y\":9,\"x\":7}"
```

The JSON RFC says "The names within an object SHOULD be unique.".

I think we should only keep the most recent one. What do you think?

Reply:
I did notice this...  Most JSON parsers are resilliant to this sort of thing.  The NPM docs even suggest using this to add comments to package.json by duplicating the key you want to comment.

I thought about only keeping the first of any value encountered (which would be the one lowest in the hierarchy of spans, since I travese the spans from bottom to top), but the downside is I'd then have to keep track of what I have and haven't seen.  I'd either need to keep a HashSet of keys or keep a HashMap of values (although then I could just serialize the whole HashMap, so it would definitely be simpler).

It's obviously "more work" from a computation perspective, because at a minimum I have to hash my keys. If I used a HashMap I'd also have to represent the values as something generic... I could use a `serde_json::Value` but it's all owned values, so I'd end up having to clone strings.  I could implement my own version of `Value` with borrowed data...  But it's probably much easier to just use a HashSet of keys.  :P

I should maybe write some kind of performance test and see how much of a difference it actually makes.  Although this is an annoying thing to test since the method signature is so complicated.  I don't even there's a public API for creating a "context".

Reply:
I have fields like `level` and `timestamp` which are `&'static str`.  I have fields from the event which are also `&'static str`.  Those are all easy.  (Although to skip fields like "level" or "timestamp", I have to copy paste a bunch of code from `tracing_serde::SerdeMapVisitor`.)

The fields from the scopes are stored in `span.extensions().get::<FormattedFields<N>>`, which are basically JSON strings for each span.  The code I liberally borrowed from to write this does:

```
let serde_json::Value::Object(fields) =
                        serde_json::from_str::<serde_json::Value>(data).unwrap()
```

to parse that into owned values, but those values are only good for that trip through the loop (and I COULD possibly parse these into not-owned-values, but I started down that rabbit hole and it's more complicated than it sounds).

I could map the set of spans to a set of Vec<Value>, and then I could use a `HashSet<Cow<str>>` to keep track of seen fields.  This is starting to get pretty complicated though.  :P

I think I'm going to leave this as a "future improvement".

I'll note that the RFC says "The names within an object SHOULD be unique," and not "MUST be unique."  I'll also note that with the implementation in tracing_subscriber, if you do:

```
    let log_to_file = tracing_subscriber::fmt::layer()
        .event_format(
            tracing_subscriber::fmt::format::json()
                .flatten_event(true)
                .with_current_span(false),
        )
        .fmt_fields(JsonFields::default());
        
    info!(level = "six", "Hello");
```

your log will have "level" repeated twice, so I have a precedent.  ;)